ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/PeerReputation.java

```diff
@@ -59,22 +63,37 @@ public PeerReputation(final int initialScore, final int maxScore) {
     checkArgument(
         initialScore <= maxScore, "Initial score must be less than or equal to max score");
     this.maxScore = maxScore;
+    this.initialScore = initialScore;
     this.score = initialScore;
   }
 
   public Optional<DisconnectReason> recordRequestTimeout(final int requestCode) {
     final int newTimeoutCount = getOrCreateTimeoutCount(requestCode).incrementAndGet();
     if (newTimeoutCount >= TIMEOUT_THRESHOLD) {
-      LOG.debug(
-          "Disconnection triggered by {} repeated timeouts for requestCode {}",
-          newTimeoutCount,
-          requestCode);
       score -= LARGE_ADJUSTMENT;
-      return Optional.of(DisconnectReason.TIMEOUT);
+      // don't trigger disconnect if this peer has a sufficiently high reputation score
+      if (peerHasNotBeenUseful()) {
+        LOG.debug(
+            "Disconnection triggered by {} repeated timeouts for requestCode {}, peer score {}",
+            newTimeoutCount,
+            requestCode,
+            score);
+        return Optional.of(DisconnectReason.TIMEOUT);
+      }
+
+      LOG.trace(
+          "Not triggering disconnect for {} repeated timeouts for requestCode {} because peer has high score {}",
```

Comment:
Should we adjust the score in this case? I might be mis-reading but if a previously useful peer becomes useless, would we never disconnect?

Reply:
the score still does get decremented on line 73

eg

```
2023-11-20 15:13:42.657+10:00 | nioEventLoopGroup-3-4 | TRACE | PeerReputation | Not triggering disconnect for exceeding useless response threshold because peer has high score 167
2023-11-20 15:13:42.716+10:00 | nioEventLoopGroup-3-2 | TRACE | PeerReputation | Not triggering disconnect for exceeding useless response threshold because peer has high score 116
2023-11-20 15:13:43.492+10:00 | nioEventLoopGroup-3-2 | DEBUG | PeerReputation | Disconnection triggered by exceeding useless response threshold, score 109
2023-11-20 14:28:19.456+10:00 | nioEventLoopGroup-3-5 | TRACE | PeerReputation | Not triggering disconnect for exceeding useless response threshold because peer has high score 117
2023-11-20 14:28:19.582+10:00 | nioEventLoopGroup-3-5 | TRACE | PeerReputation | Not triggering disconnect for exceeding useless response threshold because peer has high score 111
2023-11-20 14:28:22.063+10:00 | EthScheduler-Timer-0 | DEBUG | PeerReputation | Disconnection triggered by 13 repeated timeouts for requestCode 15, peer score 107
```